cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -624,6 +624,250 @@ TYPED_TEST(TestCaseWhenNumeric, ListOfType) {
+          DictScalarFromJSON(type, "0", dict2),
+      },
+      DictArrayFromJSON(type, "[0, 0, 2, 2]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {
+          MakeStruct({ArrayFromJSON(boolean(), "[true, true, false, false]"),
+                      ArrayFromJSON(boolean(), "[false, false, true, true]")}),
+          DictScalarFromJSON(type, "0", dict1),
+          DictScalarFromJSON(type, "0", dict2),
+      },
+      DictArrayFromJSON(type, "[0, 0, 2, 2]", dict1));
+
+  // If we can't map values from a dictionary, then raise an error
+  // Unmappable value is in the else clause
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "Cannot map dictionary index 1 at position 1 to the common dictionary"),
+      CallFunction(
+          "case_when",
+          {MakeStruct({ArrayFromJSON(boolean(), "[false, false, false, false]")}),
+           values1, values2}));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr("Cannot map dictionary index 1 to the common dictionary"),
+      CallFunction(
+          "case_when",
+          {MakeStruct({ArrayFromJSON(boolean(), "[false, false, false, false]")}),
+           values1, DictScalarFromJSON(type, "1", dict2)}));
+  // Unmappable value is in a branch (test multiple times to ensure coverage of branches
+  // in impl)
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "Cannot map dictionary index 1 at position 1 to the common dictionary"),
+      CallFunction("case_when",
+                   {MakeStruct({Datum(false),
+                                ArrayFromJSON(boolean(), "[true, true, true, true]")}),
+                    values1, values2}));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "Cannot map dictionary index 1 at position 1 to the common dictionary"),
+      CallFunction("case_when",
+                   {MakeStruct({Datum(false),
+                                ArrayFromJSON(boolean(), "[false, true, false, false]")}),
+                    values1, values2}));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr(
+          "Cannot map dictionary index 1 at position 1 to the common dictionary"),
+      CallFunction("case_when",
+                   {MakeStruct({Datum(false),
+                                ArrayFromJSON(boolean(), "[null, true, null, null]")}),
+                    values1, values2}));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      ::testing::HasSubstr("Cannot map dictionary index 1 to the common dictionary"),
+      CallFunction("case_when",
+                   {MakeStruct({Datum(false),
+                                ArrayFromJSON(boolean(), "[true, true, true, null]")}),
+                    values1, DictScalarFromJSON(type, "1", dict2)}));
+
+  // ...or optionally, emit null
+
+  // TODO: this is not implemented yet
```

Comment:
I'm not sure I understand what this TODO is for. Emitting a null when some option is enabled?